### ISSUE TITLE
Possibility to remove observable duplicates

### ISSUE THREAD

**Issue by Danko90** [jmdevince/cifpy3#6]
Hello,
I want to remove all duplicates before inserting into ES, so I should check for each observable if duplicates values exists.
This could allow some disk space to be saved.

**Comment by jmdevince** [comment]
What criteria would you want to use to remove duplicate values? Just the observable? A single observable can come from multiple sources and have multiple records for very good reasons. 

Example:
xyz.com is listed on alexa and gets added that way and tagged with whitelist. xyz.com is also in an automated blacklist and gets added that way and tagged with malware. So there will be two observables with the same value 'xyz.com' but will have come from different providers with different confidence levels and context.

Also, with ElasticSearch every day the indexes are rebuilt from scratch and old indexes are phased out. (The automatic removal isn't implemented currently in cifpy3 but it can be quickly added)

**Comment by Danko90** [comment]
Yes I agree with you, there are good reasons, depends on the using.
I need the observable value just once, that with the greater confidence value. So I thought to develop this enhancement.
Why is the old index phased out? Is it possible to save it in cifpy3?

**Comment by jmdevince** [comment]
Indexes are phased out so that old observables are removed after X days of not being seen. Like i said right now I don't have this functionality built in to this version yet. 

Yea, the idea of being able to remove duplicates while keeping the one with highest confidence isn't necessarily a bad idea. I may look at implementing that. If I do it'll have to be done *after* all feeds are ingested as a clean up routine. Querying the index at the time of inserting would be way too slow. I may use the journals to do this since the journals store a list of observables that have already been inserted from each list.

**Comment by Danko90** [comment]
I'm a little confused. After inserting of each list, the duplicates values will be removed from ES?
So, first it will be ingested and in the end removed?
Thank you

**Comment by jmdevince** [comment]
Feeder runs in hourly cycles and fetches each feed at that time. At the end of Feeder's run I can have it de-dupe objects that may have been created.

**Comment by Danko90** [comment]
Ok I got it. Another question, for each duplicate value that is removed, will lasttime value be updated?
I thought that, For example:
First time I see xyz.com and I have the firsttime.
Second time I see xyz.com again and I remove the value but the lasttime value will be updated.

Right?
What do you think?

Thanks,
Danilo

**Comment by Danko90** [comment]
Hello jmdevince,

As you requested, here an example of duplicate seen from Kibana.

![duplicates](https://cloud.githubusercontent.com/assets/1343883/10286238/1f2416c6-6b8d-11e5-8fe3-be6f5c2fa692.JPG)

Please excuse me for taking so long to answer.

**Comment by jmdevince** [comment]
I'll have to look into this.

**Comment by Danko90** [comment]
I've made some changes to the sources and managed to remove all the duplicates.
I have noticed some problem with the handling of threads, and that's why I saw duplicates from the same provider on Kibana.
Now I am checking duplicates for each list and for each observable I make a query on ES in order to see if it's already present, and the process it's not that slow.
I'll do some adjustments and within tomorrow I'll create the pull request.

Thank you very much for your explanation!

Best,
Danilo

**Comment by Danko90** [comment]
Before the pull request I'm going to create a --dedup option, in order to give the possibility to deduplicate results or not.
Also, with the --dedup option I will update the lasttime field everytime there will be a duplicate of an observable already present in ElasticSearch.
At https://github.com/Danko90/cifpy3 you can see the changes that I made until now.

Best,
Danilo

**Comment by jmdevince** [comment]
I think I'm going to revisit this in the v1.3 milestone. I plan on doing enhancements to searching and de-duplicating might be done better on the search / retrieval side rather than the input / feed consuming side.
